servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1478,6 +1519,185 @@ public void setState(long fileId, SetStateOptions options)
+      totalBlockNum = getFileBlockInfoList(fileId).size();
+
+      for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          if (workerBlockCounts.containsKey(blockLocation.workerId)) {
+            workerBlockCounts.put(blockLocation.workerId,
+                workerBlockCounts.get(blockLocation.workerId) + 1);
+          } else {
+            workerBlockCounts.put(blockLocation.workerId, 1);
+          }
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    } catch (InvalidPathException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    if (workerBlockCounts.size() == 0) {
+      LOG.error("The file " + fileId + " does not exist on any worker");
+      return IdUtils.INVALID_WORKER_ID;
+    } else if (workerBlockCounts.size() > 1) {
+      LOG.info("The file is stored at more than one worker: " + workerBlockCounts);
+    }
+
+    // return the first worker that has all the blocks
+    for (Entry<Long, Integer> entry : workerBlockCounts.entrySet()) {
+      if (entry.getValue() == totalBlockNum) {
+        return entry.getKey();
+      }
+    }
+
+    LOG.error("Not all the blocks of file {} stored on the same worker", fileId);
+    return IdUtils.INVALID_WORKER_ID;
+  }
+
+  /**
+   * Polls the files to send to the given worker for persistence.
+   *
+   * @param workerId the worker id
+   * @return the list of files
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the path is invalid
+   */
+  private List<PersistFile> pollToCheckpoint(long workerId)
+      throws FileDoesNotExistException, InvalidPathException {
+    List<PersistFile> files = Lists.newArrayList();
+    synchronized (mWorkerToAsyncPersistFiles) {
+      if (!mWorkerToAsyncPersistFiles.containsKey(workerId)) {
+        return files;
+      }
+    }
+
+    List<Long> toRequestFilePersistence = Lists.newArrayList();
+    synchronized (mWorkerToAsyncPersistFiles) {
+      Set<Long> scheduledFiles = mWorkerToAsyncPersistFiles.get(workerId);
+      for (long fileId : Sets.newHashSet(scheduledFiles)) {
+        InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+        if (inode.isCompleted()) {
+          toRequestFilePersistence.add(fileId);
+          List<Long> blockIds = Lists.newArrayList();
+          for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+            blockIds.add(fileBlockInfo.blockInfo.blockId);
+          }
+
+          PersistFile toCheckpoint = new PersistFile(fileId, blockIds);
+          files.add(toCheckpoint);
+          // update the inode file persisence state
+          inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+          scheduledFiles.remove(fileId);
```

Comment:
Hm, could we change this so that calling a method on a local reference doesn't modify external state? Maybe at the end we could call `mWorkerToAsyncPersistFiles.get(workerId).removeAll(fileIdsToPersist)`